src/Documentation/resources/Best_Practices.md

```diff
@@ -0,0 +1,422 @@
+
+-   Workload is interactive
+
+    -   Example: request-response, start/monitor/complete
+
+-   More than one server is expected or may be required\\
+
+    -   Orleans runs on a cluster which is expanded by adding servers to expand
+        the cluster
+
+-   Global coordination is not needed or on a smaller scale between a few
+    entities at a time
+
+    -   Orleans maintains its clusters and talks to other clusters. Each cluster
+        is able to communicate through tcp/ip
+
+-   *Various entities are used at different times*
+
+    -   This depends on the entities that are being used
+
+## Orleans is not the best fit when:
+
+-   Memory must be shared between entities
+
+    -   Each grain maintains its own states and should not be shared.
+
+-   A small number of large entities that may be multithreaded
+
+    -   A microservice may be a better option when supporting complex logic in a
+        single service
+
+-   Global coordination and/or consistency is needed
+
+    -   Orleans maintains the silos within its own framework
+
+-   *Operations that run for a long time*
+
+    -   Batch jobs, Single Instruction Multiple Data (SIMD) tasks
+
+    -   This depends on the need of the application and may be a fit for Orleans
+
+## Grains
+
+**Overview**:
+
+-   Actors are not objects, however they are similar
+
+-   They are loosely coupled, isolated, and primarily independent
+
+    -   Each grain is encapsulated which also maintains its own state
+        independently of other grains
+
+    -   Grains fail independently
+
+-   Avoid chatty communication between grains
+
+    -   Direct memory use is significantly less expensive than message passing
+
+    -   Highly chatty grains may be better combined as a single grain
+
+    -   Complexity/Size of arguments and serialization need to be considered
+
+        -   Deserializing twice may be more expensive than resending a binary
+            message
+
+-   Avoid bottleneck grains
+
+    -   Single coordinator/Registry/Monitor
+
+    -   Do staged aggregation If required
+
+**Asynchronicity**:
+
+-   No thread blocking: All items must be Async (Task Asynchronous Programming
+    (TAP))
+
+-   [await](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/)
+    is the best syntax to use when composing async operations
+
+-   Common Scenarios:
+
+    -   Return a concrete value:
+
+        -   return Task.FromResult(value);
+
+    -   Return a Task of the same type:
+
+        -   return foo.Bar();
+
+    -   Await a Task and continue execution:
+
+        -   `var x = await bar.Foo();  
+            var y = DoSomething(x);  
+            return y;`
+
+    -   Fan-out:
+
+        -   `var tasks = new List\<Task\>();  
+            foreach(var grain in grains)  
+            { tasks.Add(grain.Foo()) }  
+            await Task.WhenAll(tasks);  
+            DoMoreWork();`
+
+**Implementation of Grains**:
+
+-   When to use a [StatelessWorker]
```

Comment:
I would start with the first cardinal rule or grain implementation - never perform a thread-blocking operation within grain code. Examples: synchronously waiting for an IO operation or a web service call, taking a lock, running a tight loop waiting for a condition, etc. All operations other than local computations must be explicitly asynchronous.

Reply:
corrected